fuse_core/src/time.cpp

```diff
@@ -0,0 +1,136 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+
+/*
+ * This file provides a sane configuration for a clock to manage order of events.
+ * This file exists to help with a migration from fuse_core::TimeStamp to rclcpp::time
+ * Priority is given to C++ std::chrono
+ * emphasis will be placed on interoperability with fuse_core::TimeStamp and rclcpp::time
+ *
+ * This will likely be extended to distinguish between:
+ *    times of events being optimised
+ *    events generated by the running of the optimiser (log throttling, callbacks etc)
+ *
+ */
+
+#include <fuse_core/time.h>
+#include <rclcpp/utilities.hpp>
+
+using chrono_ns_time_point_t =
+  std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
+
+namespace fuse_core
+{
+TimeStamp::TimeStamp() :
+  rclcpp::Time(static_cast<int64_t>(0), RCL_CLOCK_UNINITIALIZED)
+{}
+
+TimeStamp::TimeStamp(const TimeStamp& rhs) = default;
+
+TimeStamp::TimeStamp(int32_t seconds, uint32_t nanoseconds, rcl_clock_type_t clock_type) :
+  rclcpp::Time(seconds, nanoseconds, clock_type)
+{}
+
+TimeStamp::TimeStamp(int64_t nanoseconds, rcl_clock_type_t clock_type) :
+  rclcpp::Time(nanoseconds, clock_type)
+{}
+
+TimeStamp::TimeStamp(const rclcpp::Time& time) :
+  rclcpp::Time(time)
+{}
+
+TimeStamp::TimeStamp(chrono_ns_time_point_t& stamp):
+  rclcpp::Time(stamp.time_since_epoch().count(), RCL_SYSTEM_TIME)
+{}
+
+rclcpp::Time TimeStamp::to_ros() const
+{
+  return rclcpp::Time(this->nanoseconds(), this->get_clock_type());
+}
+
+chrono_ns_time_point_t TimeStamp::to_chrono() const
+{
+  if (this->get_clock_type() != RCL_SYSTEM_TIME) {
+    throw std::runtime_error("Constructing system clock time_point from non-system clock");
+  }
+  return chrono_ns_time_point_t(
+    std::chrono::nanoseconds(this->nanoseconds())
+  );
+}
+
+
+// UTILITIES =======================================================================================
+std::ostream& operator<<(std::ostream& os, const fuse_core::TimeStamp& timestamp)
+{
+    os << timestamp.nanoseconds();
+    return os;
+}
+
+
+bool isValid(rclcpp::Clock::SharedPtr clock)
+{
+  switch (clock->get_clock_type()) {
+    case RCL_CLOCK_UNINITIALIZED:
+      return false;
+    case RCL_SYSTEM_TIME:
+      return true;
+    case RCL_ROS_TIME:
+      return clock->now().nanoseconds() == 0;
+    default:
+      return false;
+  }
+}
+
+
+// Logic adapted from: http://docs.ros.org/en/noetic/api/rostime/html/src_2time_8cpp_source.html
+bool waitForValid(rclcpp::Clock::SharedPtr clock, rclcpp::Duration timeout)
+{
+   rclcpp::Time start = clock->now();
+   while (clock->now().nanoseconds() == 0 && rclcpp::ok()) {
+     rclcpp::sleep_for(std::chrono::nanoseconds(timeout.nanoseconds()));
```

Comment:
This looks like it will unconditionally sleep for the full timeout period. If a timeout is required and we can't use `clock->sleep_until()`, I'd recommend implementing this with a jump handler on the clock itself. That way this method will return as soon as time is valid.

Reply:
I didn't realize clock->sleep_until() was available :o
I'll use that instead (well, more precisely the convenience `clock->sleep_for()`.. which I suppose uses a jump handler anyway...

Reply:
Duplicated the implementation in https://github.com/ros2/rclcpp/pull/2040 and used them https://github.com/locusrobotics/fuse/pull/283/commits/96b055546697e9c28ed6208383c4379c08fe2367